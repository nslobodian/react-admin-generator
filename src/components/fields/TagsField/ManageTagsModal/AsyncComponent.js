import React, { Component } from 'react'
import PropTypes from 'prop-types'
import connectModal from 'redux-modal/lib/connectModal'
import connect from 'react-redux/es/connect/connect'
import {
  changeNotificationTagsFilters,
  createNotificationTag,
  deleteTag,
  getNotificationPagination,
  updateTagName,
} from 'redux/NotificationTag'
import Modal from 'react-bootstrap/lib/Modal'

export class SelectProductModalWrapper extends Component {
  constructor () {
    super()

    this.state = {
      AsyncComponent: null,
    }
  }

  static propTypes = {
    show: PropTypes.bool.isRequired,
    handleHide: PropTypes.func.isRequired,
  }

  componentDidMount () {
    require.ensure(
      [],
      require => this.setState({ AsyncComponent: require('./ManageTagsModal').default }),
      'ManageTagsModal'
    )
  }

  updateTagNameHandler = (tagId) => () => this.props.updateTagName(tagId)

  render () {
    const {
      show, handleHide, createNotificationTag, notificationTags,
      getNotificationPagination, pagination, deleteTag, changeNotificationTagsFilters,
    } = this.props
    const { AsyncComponent } = this.state
    const initialValues = {}
    notificationTags.forEach(({ name, notificationTagId }) => { initialValues[notificationTagId] = name })
    return (
      <Modal show={show}
        onHide={handleHide}
        bsSize='sm'
        className='onlyCloseButton'>
        <Modal.Header closeButton />
        <Modal.Body>
          {
            AsyncComponent &&
            <AsyncComponent
              createNotificationTag={createNotificationTag}
              notificationTags={notificationTags}
              initialValues={initialValues}
              deleteTag={deleteTag}
              updateTagName={this.updateTagNameHandler}
              onPageChange={getNotificationPagination}
              pagination={pagination}
              filterChangeHandler={changeNotificationTagsFilters}
            />
          }
        </Modal.Body>
      </Modal>
    )
  }
}

const mapActionCreators = {
  createNotificationTag,
  deleteTag,
  getNotificationPagination,
  updateTagName,
  changeNotificationTagsFilters,
}

export default connect(({ NotificationTag: { notificationTags, pagination } }) => ({
  notificationTags,
  pagination,
}), mapActionCreators)(
  connectModal({ name: 'mangeTagsModal', destroyOnHide: true })(SelectProductModalWrapper)
)
