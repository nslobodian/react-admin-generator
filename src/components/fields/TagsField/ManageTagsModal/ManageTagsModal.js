import React from 'react'
import reduxForm from 'redux-form/es/reduxForm'
import Field from 'redux-form/es/Field'
import s from './ManageTagsModal.scss'
import withState from 'recompose/withState'
import withHandlers from 'recompose/withHandlers'
import { textHandler } from 'utils/onChangeFilterHandlers'
import Pagination from 'react-js-pagination'
import DeleteButtonWithConfirmation from 'components/DeleteButtonWithConfirmation'

export const EditTagNameField = withState('open', 'toggleStatus', false)(withHandlers({
  toggleStatus: ({ toggleStatus }) => () => {
    toggleStatus(n => !n)
  },
})(({ name, notificationTagId, deleteTag, toggleStatus, open, updateTagName }) =>
  <tr>
    <td>
      {open
        ? <Field
          component='input'
          type='text'
          className='form-control'
          name={notificationTagId}
        />
        : name
      }
    </td>
    <td className='EditButton' onClick={toggleStatus}>
      {open
        ? <div onClick={updateTagName}>Save</div>
        : <div>Edit</div>
      }
    </td>
    <td><DeleteButtonWithConfirmation onDelete={deleteTag} deletingId={notificationTagId} item='tag' /></td>
  </tr>
))

export const ManageTagsModal = ({
  createNotificationTag, notificationTags, deleteTag,
  updateTagName, onPageChange, pagination: { page, perPage, totalCount },
  filterChangeHandler,
}) =>
  <div className={s.ManageTagsModal}>
    <div className={s.modalHeader}>Manage Tags</div>
    <div>
      <div className={s.NewTagWrapper}>
        <div className='text-muted'>New Tag</div>
        <Field
          component='input'
          type='text'
          className='form-control'
          name='newTag'
        />
      </div>
      <button className='btn btn-primary' type='button' onClick={createNotificationTag}>Add</button>
    </div>
    <div className={s.TagsTableWrapper}>
      <table className='table'>
        <thead>
          <tr>
            <td colSpan='3'>Tags</td>
          </tr>
          <tr>
            <td colSpan='3'>
              <Field
                component='input'
                type='text'
                className='form-control'
                placeholder='Filter'
                name='tagName'
                onChange={textHandler(filterChangeHandler, 'name')}
              />
            </td>
          </tr>
        </thead>
        <tbody>
          {notificationTags.length > 0
            ? notificationTags.map(({ name, notificationTagId }, index) =>
              <EditTagNameField
                key={index}
                name={name}
                notificationTagId={notificationTagId}
                updateTagName={updateTagName(notificationTagId)}
              />
            )
            : <tr>
              <td colSpan='3' className='text-center'>No Tags</td>
            </tr>
          }
          <tr>
            <td colSpan='3' className='text-center'>
              <Pagination
                activePage={page}
                itemsCountPerPage={perPage}
                totalItemsCount={totalCount || 0}
                pageRangeDisplayed={5}
                onChange={onPageChange}
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>

export default reduxForm({
  form: 'ManageTagsForm',
})(ManageTagsModal)
