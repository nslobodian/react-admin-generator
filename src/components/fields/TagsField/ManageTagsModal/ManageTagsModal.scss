.ManageTagsModal {
  .modalHeader {
    margin: 20px 0;
    font-size: 18px;
    text-align: center;
  }

  .NewTagWrapper {
    width: 70%;
    display: inline-block;
    margin-right: 10px;
    margin-bottom: 10px;
  }

  .TagsTableWrapper {
    td + td {
      width: 30px;
    }

    .name {
      padding-left: 10px;
      margin-bottom: 0;
      font-weight: 400;
      cursor: pointer;
    }
  }
}