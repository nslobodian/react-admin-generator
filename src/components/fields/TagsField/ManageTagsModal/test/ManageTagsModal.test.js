import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { EditTagNameField, ManageTagsModal } from '../ManageTagsModal'

describe('(Components) ManageTagsModal', () => {
  it('Should render component properly', () => {
    const props = {
      pagination: {
        page: 1,
        perPage: 10,
        totalCount: 1,
      },
      filterChangeHandler: jest.fn(),
      onPageChange: jest.fn(),
      updateTagName: jest.fn(),
      deleteTag: jest.fn(),
      notificationTags: [
        {
          name: 'name',
          notificationTagId: 1,
        },
      ],
      createNotificationTag: jest.fn(),
    }
    const wrapper = shallow(
      <ManageTagsModal
        {...props}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  }),
  it('Without notificationTags', () => {
    const props = {
      pagination: {
        page: 1,
        perPage: 10,
      },
      filterChangeHandler: jest.fn(),
      onPageChange: jest.fn(),
      updateTagName: jest.fn(),
      deleteTag: jest.fn(),
      notificationTags: {},
      createNotificationTag: jest.fn(),
    }
    const wrapper = shallow(
      <ManageTagsModal
        {...props}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  }),
  it('EditTagNameField', () => {
    const props = {
      name: 'name',
      notificationTagId: 1,
      deleteTag: jest.fn(),
      toggleStatus: jest.fn(),
      open: true,
      toggleStatus: jest.fn(),
      updateTagName: jest.fn(),
    }
    const wrapper = shallow(
      <EditTagNameField
        {...props}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
