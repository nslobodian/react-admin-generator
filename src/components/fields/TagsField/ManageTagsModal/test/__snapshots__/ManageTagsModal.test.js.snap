// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Components) ManageTagsModal EditTagNameField 1`] = `
<_class
  deleteTag={[Function]}
  name="name"
  notificationTagId={1}
  open={false}
  toggleStatus={[Function]}
  updateTagName={[Function]}
/>
`;

exports[`(Components) ManageTagsModal Should render component properly 1`] = `
<div
  className="ManageTagsModal"
>
  <div
    className="modalHeader"
  >
    Manage Tags
  </div>
  <div>
    <div
      className="NewTagWrapper"
    >
      <div
        className="text-muted"
      >
        New Tag
      </div>
      <Field
        className="form-control"
        component="input"
        name="newTag"
        type="text"
      />
    </div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
  <div
    className="TagsTableWrapper"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <td
            colSpan="3"
          >
            Tags
          </td>
        </tr>
        <tr>
          <td
            colSpan="3"
          >
            <Field
              className="form-control"
              component="input"
              name="tagName"
              onChange={[Function]}
              placeholder="Filter"
              type="text"
            />
          </td>
        </tr>
      </thead>
      <tbody>
        <_class2
          name="name"
          notificationTagId={1}
        />
        <tr>
          <td
            className="text-center"
            colSpan="3"
          >
            <Pagination
              activePage={1}
              firstPageText="«"
              innerClass="pagination"
              itemsCountPerPage={10}
              lastPageText="»"
              nextPageText="⟩"
              onChange={[Function]}
              pageRangeDisplayed={5}
              prevPageText="⟨"
              totalItemsCount={1}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`(Components) ManageTagsModal Without notificationTags 1`] = `
<div
  className="ManageTagsModal"
>
  <div
    className="modalHeader"
  >
    Manage Tags
  </div>
  <div>
    <div
      className="NewTagWrapper"
    >
      <div
        className="text-muted"
      >
        New Tag
      </div>
      <Field
        className="form-control"
        component="input"
        name="newTag"
        type="text"
      />
    </div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
  <div
    className="TagsTableWrapper"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <td
            colSpan="3"
          >
            Tags
          </td>
        </tr>
        <tr>
          <td
            colSpan="3"
          >
            <Field
              className="form-control"
              component="input"
              name="tagName"
              onChange={[Function]}
              placeholder="Filter"
              type="text"
            />
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="text-center"
            colSpan="3"
          >
            No Tags
          </td>
        </tr>
        <tr>
          <td
            className="text-center"
            colSpan="3"
          >
            <Pagination
              activePage={1}
              firstPageText="«"
              innerClass="pagination"
              itemsCountPerPage={10}
              lastPageText="»"
              nextPageText="⟩"
              onChange={[Function]}
              pageRangeDisplayed={5}
              prevPageText="⟨"
              totalItemsCount={0}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
